### PR TITLE
ngraph-1676 batch dot fusion

### PR DESCRIPTION
**Pull request by louisfeng**
BatchDot fuses multiple dot operations.

### CODE REVIEW

File: src/ngraph/runtime/cpu/op/batch_dot.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    NGRAPH_DEBUG << "dot_dimension_a = " << dot_dimension_a
+                 << " , dot_dimension_b = " << dot_dimension_b;
+    NGRAPH_DEBUG << "a shape = " << vector_to_string(shape_a)
+                 << " , b shape = " << vector_to_string(shape_b);
+
+    if (shape_a.at(dot_dimension_a) != shape_b.at(dot_dimension_b))
+    {
+        throw ngraph_error("product dimensions are not equal while creating BatchDot");
+    }
+
```
Comment (pruthviIntel) [inline src/ngraph/runtime/cpu/op/batch_dot.cpp]: Can we also add the check, if the element type for a and b are same?

Comment (louisfeng) [reply src/ngraph/runtime/cpu/op/batch_dot.cpp]: Good point, I'll add that.

File: src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp [hunk 2]
```diff
@@ -199,3 +201,104 @@ bool runtime::cpu::pass::CPURnnMatFusion::run_on_function(std::shared_ptr<Functi
     }
     return modified;
 }
+
+#define TI(x) std::type_index(typeid(x))
+
+std::shared_ptr<Node> fuse_batch_dot(const std::shared_ptr<Node>& n)
+{
+    auto reshape_pred = [](const std::shared_ptr<Node>& r) {
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp]: could we please use `has_class` to reduce the boiler-plate?

File: src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp [hunk 3]
```diff
@@ -199,3 +201,104 @@ bool runtime::cpu::pass::CPURnnMatFusion::run_on_function(std::shared_ptr<Functi
...
+    return {nullptr};
+}
+
+bool runtime::cpu::pass::CPUBatchDotFusion::run_on_function(std::shared_ptr<Function> func)
+{
+    bool modified = false;
+
+    for (auto n : func->get_ordered_ops())
+    {
+        const Node& node = *n;
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp]: if we wud just do `std::dynamic_pointer_cast<op::Concat>` we could avoid all this `*n` and `TI` class comparisons 😃

File: src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp [hunk 4]
```diff
@@ -199,3 +201,104 @@ bool runtime::cpu::pass::CPURnnMatFusion::run_on_function(std::shared_ptr<Functi
...
+                }
+            }
+        }
+    }
+    if (fuse_input[0] && fuse_input[1])
+    {
+        return std::make_shared<op::BatchDot>(
+            fuse_input[0], fuse_input[1], transpose[0], transpose[1]);
+    }
+    return {nullptr};
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp]: what do we need the brackets for?

File: src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp [hunk 5]
```diff
@@ -199,3 +201,104 @@ bool runtime::cpu::pass::CPURnnMatFusion::run_on_function(std::shared_ptr<Functi
...
+                        ++reshape_count[i];
+                        if (reshape_count[i] == num_expected_reshape_with_trans)
+                        {
+                            transpose[i] = true;
+                            break;
+                        }
+                    }
+                    // branch to either input 0 or 1 depending on which one we are traversing
+                    iter =
+                        iter->get_input_size() > 1 ? iter->get_argument(i) : iter->get_argument(0);
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp]: this is cool  😄

File: src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp [hunk 6]
```diff
@@ -199,3 +201,104 @@ bool runtime::cpu::pass::CPURnnMatFusion::run_on_function(std::shared_ptr<Functi
...
+            // we found a match, determine whether we have to transpose for each input by
+            // counting the number of reshapes in each branch, if transpose is applied, there
+            // should be 3 reshapes.
+            for (int i = 0; i < num_op_branches; ++i)
+            {
+                auto iter = matcher->get_match_root();
+                auto& input_node = pattern_map[input[i]];
+                do
+                {
+                    if (std::dynamic_pointer_cast<op::Reshape>(iter) != nullptr)
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp]: do we want to check what those reshapes actually do (i.e. transpose or reshape or god forbids both at the same time) ?

File: src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp (hunk 4 above)
Comment (louisfeng) [reply src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp]: Return type is `shared_ptr`

File: src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp (hunk 3 above)
Comment (louisfeng) [reply src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp]: Copy&paste our conversation: Dynamic cast actually works for derived class too. So there's potentially ambiguity in this check and we want `concat` exactly, not potentially a derived class of `concat`

**Pull request merged**